<a href="https://colab.research.google.com/github/Valeria379/Data-Science/blob/main/%D0%9C%D0%BE%D1%80%D0%BE%D0%B7%D0%BE%D0%B2%D0%B0_%D0%92_04_2_HW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Практическая работа 4-2: Линейная регрессия 2

Name: Морозова Валерия

Email: Morozova-VA@mgpu.ru


## ##  Анализ набора кредитных данных

Вспомните набор данных «Кредит», рассмотренный на занятии и доступный [здесь](http://www-bcf.usc.edu/~gareth/ISL/data.html).
Этот набор данных состоит из информации о кредитных картах 400 человек.

Сначала импортируйте данные и преобразуйте доход в тысячи.

In [1]:
from google.colab import files

uploaded = files.upload()

Saving Credit.csv to Credit.csv


In [2]:
# imports and setup

import scipy as sc
import numpy as np

import pandas as pd
import statsmodels.formula.api as sm     #Last lecture: used statsmodels.formula.api.ols() for OLS
from sklearn import linear_model         #Last lecture: used sklearn.linear_model.LinearRegression() for OLS

import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (10, 6)

from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm

# Import data from Credit.csv file
credit = pd.read_csv('Credit.csv',index_col=0) #load data
credit["Income"] = credit["Income"].map(lambda x: 1000*x)
credit

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1     14891.0   3606     283      2   34         11    Male      No     Yes   
2    106025.0   6645     483      3   82         15  Female     Yes     Yes   
3    104593.0   7075     514      4   71         11    Male      No      No   
4    148924.0   9504     681      3   36         11  Female      No      No   
5     55882.0   4897     357      2   68         16    Male      No     Yes   
..        ...    ...     ...    ...  ...        ...     ...     ...     ...   
396   12096.0   4100     307      3   32         13    Male      No     Yes   
397   13364.0   3838     296      5   65         17    Male      No      No   
398   57872.0   4171     321      5   67         12  Female      No     Yes   
399   37728.0   2525     192      1   44         13    Male      No     Yes   
400   18701.0   5524     415      5   64          7  Female      No      No   

            Ethnicity  Balance  
1           Caucasian      333  
2               Asian      903  
3               Asian      580  
4               Asian      964  
5           Caucasian      331  
..                ...      ...  
396         Caucasian      560  
397  African American      480  
398         Caucasian      138  
399         Caucasian        0  
400             Asian      966  

[400 rows x 11 columns]

In [3]:
credit.shape

(400, 11)

In [4]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
Index: 400 entries, 1 to 400
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Income     400 non-null    float64
 1   Limit      400 non-null    int64  
 2   Rating     400 non-null    int64  
 3   Cards      400 non-null    int64  
 4   Age        400 non-null    int64  
 5   Education  400 non-null    int64  
 6   Gender     400 non-null    object 
 7   Student    400 non-null    object 
 8   Married    400 non-null    object 
 9   Ethnicity  400 non-null    object 
 10  Balance    400 non-null    int64  
dtypes: float64(1), int64(6), object(4)
memory usage: 37.5+ KB


In [5]:
print(credit.describe())

              Income         Limit      Rating       Cards         Age  \
count     400.000000    400.000000  400.000000  400.000000  400.000000   
mean    45218.885000   4735.600000  354.940000    2.957500   55.667500   
std     35244.273165   2308.198848  154.724143    1.371275   17.249807   
min     10354.000000    855.000000   93.000000    1.000000   23.000000   
25%     21007.250000   3088.000000  247.250000    2.000000   41.750000   
50%     33115.500000   4622.500000  344.000000    3.000000   56.000000   
75%     57470.750000   5872.750000  437.250000    4.000000   70.000000   
max    186634.000000  13913.000000  982.000000    9.000000   98.000000   

        Education      Balance  
count  400.000000   400.000000  
mean    13.450000   520.015000  
std      3.125207   459.758877  
min      5.000000     0.000000  
25%     11.000000    68.750000  
50%     14.000000   459.500000  
75%     16.000000   863.000000  
max     20.000000  1999.000000  


## Задание 1. Первая регрессионная модель

**Упражнение:** Лимит средств по рейтингу:


$$
\text{Limit} = \beta_0 + \beta_1 \text{Rating}.
$$

Поскольку кредитные рейтинги в основном используются банками для определения кредитных лимитов, ожидаем, что рейтинг будет в прямой зависимости для лимита, поэтому эта регрессия должна быть очень хорошей.

Используйте функцию `ols` из библиотеки Python `statsmodels`. Каково значение $R^2$? Что такое $H_0$ и $H_A$ для соответствующей проверки гипотезы и каково значение $p$?


$H_0$ (нулевая гипотеза) - кредитный рейтинг никак не влияет на определения кредитного лимита

$H_A$ (альтернативная гипотеза) - кредитный лимит находится в прямой зависимости от кредитного рейтинга

In [6]:
# your code goes here
model = sm.ols(formula="Limit ~ Rating", data=credit).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.994
Model:                            OLS   Adj. R-squared:                  0.994
Method:                 Least Squares   F-statistic:                 6.348e+04
Date:                Tue, 23 Apr 2024   Prob (F-statistic):               0.00
Time:                        10:39:37   Log-Likelihood:                -2649.1
No. Observations:                 400   AIC:                             5302.
Df Residuals:                     398   BIC:                             5310.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -542.9282     22.850    -23.760      0.000    -587.851    -498.006
Rating        14.8716      0.059    251.949      0.000      14.756      14.988
==============================================================================
Omnibus:                        6.887   Durbin-Watson:                   2.080
Prob(Omnibus):                  0.032   Jarque-Bera (JB):                4.980
Skew:                          -0.145   Prob(JB):                       0.0829
Kurtosis:                       2.537   Cond. No.                         970.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Ваш ответ:**

$\hat{\beta}_0  = -542.93$

$\hat{\beta}_1  = 14.87$

**Уравнение получается**: Limit = -542.93 + 14.87*Rating

**$R^2$**= 0.994% это значит,что модель объясняет $99.4\%$ изменчивости кредитного лимита. Это достаточно высокое значение, указывающее на хорошее соответствие модели данным.

Значение **$p$** крайне мало (=0.000 < 0.05), значит, отвергаем нулевую гипотезу и говорим, что эта регрессия демонстрирует сильную линейную связь и подтверждает значимость модели, и лимит находится в прямой зависимости от рейтинга.

## Задание 2: Прогнозирование лимита без рейтинга

Поскольку `Рейтинг` и `Лимит` — это практически одна и та же переменная, исключим `Рейтинг` и просто попробуем спрогнозировать `Лимит` на основе вещественнозначных переменных (некатегориальных переменных): Доход, Карты, Возраст, Образование, Баланс.

**Упражнение.** Разработайте модель множественной линейной регрессии для прогнозирования лимита. Интерпретируйте результаты.

А пока просто сосредоточьтесь на переменных с численными значениями (доход, карты, возраст, образование, баланс).

В первом приближении не использовать категориальные переменные (Пол, Студент, Семейное положение, Этническая принадлежность).



In [7]:
#Разработаем модель множественной линейной регрессии для некатегориальных переменных
model_2 = sm.ols(formula="Limit ~ Income + Cards + Age + Education + Balance", data=credit).fit()
model_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.940
Model:                            OLS   Adj. R-squared:                  0.940
Method:                 Least Squares   F-statistic:                     1242.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          1.32e-238
Time:                        10:39:44   Log-Likelihood:                -3101.0
No. Observations:                 400   AIC:                             6214.
Df Residuals:                     394   BIC:                             6238.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1752.3883    170.621     10.271      0.000    1416.947    2087.830
Income         0.0323      0.001     34.737      0.000       0.031       0.034
Cards        -61.7552     20.904     -2.954      0.003    -102.852     -20.658
Age            1.9710      1.683      1.171      0.242      -1.337       5.279
Education     -4.9297      9.107     -0.541      0.589     -22.834      12.974
Balance        3.1921      0.070     45.291      0.000       3.053       3.331
==============================================================================
Omnibus:                       78.568   Durbin-Watson:                   1.992
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              124.063
Skew:                          -1.356   Prob(JB):                     1.15e-27
Kurtosis:                       3.306   Cond. No.                     3.45e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.45e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

По предварительной оценке, можно сказать, что хорошими предикторами являются переменные Income и Balance, так как их p-значение значительно меньше 0.05. Чтобы точно убедиться, есть ли среди других переменных хорошие или они будут считаться плохими, разработаем отдельные модели линейной регрессии.

In [8]:
model_2_cards = sm.ols(formula="Limit ~ Cards", data=credit).fit()
model_2_cards.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                   0.04167
Date:                Tue, 23 Apr 2024   Prob (F-statistic):              0.838
Time:                        10:40:06   Log-Likelihood:                -3664.7
No. Observations:                 400   AIC:                             7333.
Df Residuals:                     398   BIC:                             7341.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4684.6662    274.978     17.037      0.000    4144.076    5225.257
Cards         17.2219     84.369      0.204      0.838    -148.643     183.087
==============================================================================
Omnibus:                       46.466   Durbin-Watson:                   1.959
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               62.069
Skew:                           0.833   Prob(JB):                     3.33e-14
Kurtosis:                       3.973   Cond. No.                         8.37
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

$R^2$ у значения Cards = 0.000, а p-значение гораздо больше 0.05, значит модель не значима и связи между переменными нет.

In [9]:
model_2_age = sm.ols(formula="Limit ~ Age", data=credit).fit()
model_2_age.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.010
Model:                            OLS   Adj. R-squared:                  0.008
Method:                 Least Squares   F-statistic:                     4.093
Date:                Tue, 23 Apr 2024   Prob (F-statistic):             0.0437
Time:                        10:40:18   Log-Likelihood:                -3662.7
No. Observations:                 400   AIC:                             7329.
Df Residuals:                     398   BIC:                             7337.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3984.0985    388.857     10.246      0.000    3219.629    4748.568
Age           13.4998      6.673      2.023      0.044       0.381      26.619
==============================================================================
Omnibus:                       41.937   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               54.407
Skew:                           0.782   Prob(JB):                     1.53e-12
Kurtosis:                       3.906   Cond. No.                         197.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

$R^2$ у значения Age = 0.010, а p-значение хоть и меньше 0.05, но близко к нему, значит модель объясняет всего $1\%$ изменчивости кредитного лимита в зависимости от возраста.

In [10]:
model_2_education = sm.ols(formula="Limit ~ Education", data=credit).fit()
model_2_education.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.2208
Date:                Tue, 23 Apr 2024   Prob (F-statistic):              0.639
Time:                        10:40:29   Log-Likelihood:                -3664.7
No. Observations:                 400   AIC:                             7333.
Df Residuals:                     398   BIC:                             7341.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4969.5271    511.029      9.725      0.000    3964.874    5974.180
Education    -17.3924     37.011     -0.470      0.639     -90.154      55.369
==============================================================================
Omnibus:                       47.086   Durbin-Watson:                   1.956
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               63.382
Skew:                           0.837   Prob(JB):                     1.72e-14
Kurtosis:                       3.999   Cond. No.                         61.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

$R^2$ у значения Education = 0.001, а p-значение гораздо больше 0.05 < (0.639), значит модель не значима и связи между переменными нет.

In [11]:
model_2_balance = sm.ols(formula="Limit ~ Balance", data=credit).fit()
model_2_balance.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.743
Model:                            OLS   Adj. R-squared:                  0.742
Method:                 Least Squares   F-statistic:                     1148.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          2.53e-119
Time:                        10:40:39   Log-Likelihood:                -3393.4
No. Observations:                 400   AIC:                             6791.
Df Residuals:                     398   BIC:                             6799.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2485.9566     88.586     28.063      0.000    2311.802    2660.111
Balance        4.3261      0.128     33.879      0.000       4.075       4.577
==============================================================================
Omnibus:                       21.227   Durbin-Watson:                   1.967
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               29.462
Skew:                           0.424   Prob(JB):                     4.00e-07
Kurtosis:                       4.024   Cond. No.                     1.05e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.05e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**$R^2$**= 0.743% у значения Balance, это значит,что модель объясняет $74.3\%$ изменчивости кредитного лимита. P-значение меньше 0.05, всё указывает на хорошее соответствие модели данным.

In [12]:
model_2_income = sm.ols(formula="Limit ~ Income", data=credit).fit()
model_2_income.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.627
Model:                            OLS   Adj. R-squared:                  0.626
Method:                 Least Squares   F-statistic:                     670.2
Date:                Tue, 23 Apr 2024   Prob (F-statistic):           2.39e-87
Time:                        10:42:32   Log-Likelihood:                -3467.3
No. Observations:                 400   AIC:                             6939.
Df Residuals:                     398   BIC:                             6947.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2389.8694    114.829     20.812      0.000    2164.123    2615.616
Income         0.0519      0.002     25.888      0.000       0.048       0.056
==============================================================================
Omnibus:                      141.700   Durbin-Watson:                   1.966
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               21.008
Skew:                          -0.064   Prob(JB):                     2.74e-05
Kurtosis:                       1.885   Cond. No.                     9.33e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.33e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**$R^2$**= 0.627% у значения Income, это значит,что модель объясняет $62.7\%$ изменчивости кредитного лимита. P-значение меньше 0.05, всё указывает на неплохое соответствие модели данным.

In [13]:
best_model = sm.ols(formula="Limit ~ Income + Balance", data=credit).fit()
best_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.939
Model:                            OLS   Adj. R-squared:                  0.939
Method:                 Least Squares   F-statistic:                     3046.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          1.37e-241
Time:                        10:42:44   Log-Likelihood:                -3106.0
No. Observations:                 400   AIC:                             6218.
Df Residuals:                     397   BIC:                             6230.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1610.3665     49.710     32.395      0.000    1512.639    1708.094
Income         0.0327      0.001     35.691      0.000       0.031       0.035
Balance        3.1621      0.070     44.955      0.000       3.024       3.300
==============================================================================
Omnibus:                       72.111   Durbin-Watson:                   1.951
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              109.872
Skew:                          -1.278   Prob(JB):                     1.39e-24
Kurtosis:                       3.241   Cond. No.                     9.95e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.95e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Какие независимые переменные являются хорошими/плохими предикторами? Какая модель в целом является лучшей?

**Ваш ответ:**

На основе проведенного анализа можно c уверенностью сказать, что:

Плохими предиторами являются переменные: Cards, Age, Education

Хорошие предиторами являются переменные: Income и Balance, из которых получается лучшая модель.

$$
\text{Limit} = \beta_0 + \beta_1 * \text{Income} + \beta_2*\text{Balance}
$$

$\hat{\beta}_0  = 1610.37$

$\hat{\beta}_1  = 0.03$

$\hat{\beta}_2  = 3.16$

**Уравнение получается**: $$
\text{Limit} = 1610.37 + 0.03 * \text{Income} + 3.16*\text{Balance}
$$

**$R^2$**= 0.939% это значит,что модель объясняет $93.9\%$ изменчивости кредитного лимита. Это достаточно высокое значение, указывающее на хорошее соответствие модели данным.

Значение **$p$** мало (=0.000 < 0.05), значит, что эта регрессия демонстрирует сильную линейную связь и подтверждает значимость модели.

## Кросс-валидация

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# Есть данные в виде массивов numpy: X - матрица признаков (оказывающих большее влияние на кредитный лимит), y - вектор целевой переменной (кредитный лимит)
charges_encoded = pd.get_dummies(credit, columns=['Limit'])
X = charges_encoded[['Income', 'Balance']]
y = credit['Limit']

# Разделим данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Обучим регрессионную модель
model = LinearRegression()
model.fit(X_train, y_train)

# Получим предсказания для тестового набора
y_pred = model.predict(X_test)

# Рассчитаем среднеквадратичную ошибку (RMSE)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Рассчитаем RMSE в процентах
y_mean = np.mean(y_test)
rmse_percent = (rmse / y_mean) * 100

print(f'Среднеквадратичная ошибка (RMSE): {rmse}')
print(f'Среднеквадратичная ошибка (RMSE) в процентах: {rmse_percent:.2f}%')

Среднеквадратичная ошибка (RMSE): 577.6564241806103
Среднеквадратичная ошибка (RMSE) в процентах: 12.08%


Чем ниже RMSE, тем лучше данная модель может «соответствовать» набору данных. Если оценивать результат в процентах (от 0 до 100), то можно сказать, что данная модель хорошо соответствует набору данных и в дальнейшем будет успешно *обобщаться* на независимом наборе данных,
и прогнозировать кредитный лимит для параметров, которые не использовались для *обучения* модели.

##Визуализация

In [15]:
import matplotlib.pyplot as plt
import plotly.express as px
%matplotlib inline
plt.rcParams['figure.figsize'] = (10, 6)

from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm

In [16]:
fig = px.scatter(credit,
                 x='Income',
                 y='Limit',
                 color='Gender',
                 opacity=0.8,
                 hover_data=['Gender'],
                 title='Income vs Limit',)
fig.update_traces(marker_size=5)
fig.show()

График рассеивания наглядно демонстрирует и подтвержает: чем больше доход, тем больше будет кредитный лимит, независимо от полового признака.

## Задание 3. Включение категориальных переменных в модель регрессии

Рассмотреть двоичные категориальные переменные, которые были ранее сопоставлены с целочисленными значениями 0, 1 в в `лекции 4.2`.

In [17]:
credit["Gender_num"] = credit["Gender"].map({' Male':0, 'Female':1})
credit["Student_num"] = credit["Student"].map({'Yes':1, 'No':0})
credit["Married_num"] = credit["Married"].map({'Yes':1, 'No':0})

Можете ли улучшить модель, которая разработана в задании `2`, включив в нее одну или несколько из этих переменных?


In [18]:
credit


Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1     14891.0   3606     283      2   34         11    Male      No     Yes   
2    106025.0   6645     483      3   82         15  Female     Yes     Yes   
3    104593.0   7075     514      4   71         11    Male      No      No   
4    148924.0   9504     681      3   36         11  Female      No      No   
5     55882.0   4897     357      2   68         16    Male      No     Yes   
..        ...    ...     ...    ...  ...        ...     ...     ...     ...   
396   12096.0   4100     307      3   32         13    Male      No     Yes   
397   13364.0   3838     296      5   65         17    Male      No      No   
398   57872.0   4171     321      5   67         12  Female      No     Yes   
399   37728.0   2525     192      1   44         13    Male      No     Yes   
400   18701.0   5524     415      5   64          7  Female      No      No   

            Ethnicity  Balance  Gender_num  Student_num  Married_num  
1           Caucasian      333           0            0            1  
2               Asian      903           1            1            1  
3               Asian      580           0            0            0  
4               Asian      964           1            0            0  
5           Caucasian      331           0            0            1  
..                ...      ...         ...          ...          ...  
396         Caucasian      560           0            0            1  
397  African American      480           0            0            0  
398         Caucasian      138           1            0            1  
399         Caucasian        0           0            0            1  
400             Asian      966           1            0            0  

[400 rows x 14 columns]

Изучим модель, к которой добавятся все каьегориальные данные.

In [19]:
new_model = sm.ols(formula="Limit ~ Income + Balance + Gender_num + Student_num + Married_num", data=credit).fit()
new_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.974
Model:                            OLS   Adj. R-squared:                  0.973
Method:                 Least Squares   F-statistic:                     2924.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          7.24e-309
Time:                        11:17:53   Log-Likelihood:                -2936.7
No. Observations:                 400   AIC:                             5885.
Df Residuals:                     394   BIC:                             5909.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    1638.9094     44.483     36.844      0.000    1551.457    1726.362
Income          0.0311      0.001     51.197      0.000       0.030       0.032
Balance         3.4721      0.048     72.020      0.000       3.377       3.567
Gender_num     47.6680     37.723      1.264      0.207     -26.496     121.832
Student_num -1497.0609     65.646    -22.805      0.000   -1626.121   -1368.001
Married_num    14.2441     38.769      0.367      0.714     -61.975      90.464
==============================================================================
Omnibus:                       86.931   Durbin-Watson:                   1.886
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              144.469
Skew:                          -1.301   Prob(JB):                     4.26e-32
Kurtosis:                       4.378   Cond. No.                     2.01e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.01e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Исходя из значения **$R^2$**= 0.974%, можно сказать, что при включении категориальных переменных модель улучшилась, однако согласно p-значению видно, что Gender_num и Married_num значительно превышают норму (0.05), поэтому их можно убрать и сказать, что половой признак и семейное положение не влияют на кредитный лимит.

In [20]:
new_model_best = sm.ols(formula="Limit ~ Income + Balance + Student_num", data=credit).fit()
new_model_best.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Limit   R-squared:                       0.974
Model:                            OLS   Adj. R-squared:                  0.973
Method:                 Least Squares   F-statistic:                     4875.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          3.55e-312
Time:                        11:17:58   Log-Likelihood:                -2937.6
No. Observations:                 400   AIC:                             5883.
Df Residuals:                     396   BIC:                             5899.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    1671.8940     32.784     50.998      0.000    1607.442    1736.346
Income          0.0311      0.001     51.242      0.000       0.030       0.032
Balance         3.4730      0.048     72.069      0.000       3.378       3.568
Student_num -1494.7953     65.365    -22.868      0.000   -1623.301   -1366.290
==============================================================================
Omnibus:                       86.280   Durbin-Watson:                   1.885
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              142.706
Skew:                          -1.295   Prob(JB):                     1.03e-31
Kurtosis:                       4.360   Cond. No.                     1.99e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.99e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Значение **$R^2$**= 0.974% не изменилось после того, как были убраны два незначимых параметра, можно также сделать вывод, что если человек является студентом, то его кредитный лимит будет значительно меньше, чем у тех, кто закончил обучение и имеет высокий доход. Чтобы убедиться в том, что данная модель максимально усовершенствована, проведем кросс-валидацию (перекрестную проверку).

## Кросс-валидация

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# Есть данные в виде массивов numpy: X - матрица признаков (оказывающих большее влияние на кредитный лимит), y - вектор целевой переменной (кредитный лимит)
charges_encoded = pd.get_dummies(credit, columns=['Limit'])
X = charges_encoded[['Income', 'Balance', 'Student_num']]
y = credit['Limit']

# Разделим данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Обучим регрессионную модель
model = LinearRegression()
model.fit(X_train, y_train)

# Получим предсказания для тестового набора
y_pred = model.predict(X_test)

# Рассчитаем среднеквадратичную ошибку (RMSE)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Рассчитаем RMSE в процентах
y_mean = np.mean(y_test)
rmse_percent = (rmse / y_mean) * 100

print(f'Среднеквадратичная ошибка (RMSE): {rmse}')
print(f'Среднеквадратичная ошибка (RMSE) в процентах: {rmse_percent:.2f}%')

Среднеквадратичная ошибка (RMSE): 318.03785858280656
Среднеквадратичная ошибка (RMSE) в процентах: 6.65%


**Ваш ответ:**

Данная модель может «соответствовать» набору данных. Если оценивать результат в процентах (от 0 до 100), то можно сказать, что Среднеквадратичная ошибка сократилась в 2 раза, данная модель отлично соответствует набору данных и в дальнейшем будет успешно *обобщаться* на независимом наборе,
и прогнозировать зависимое значение для параметров, которые не использовались для *обучения* модели.

##Визуализация

In [26]:
fig = px.scatter(credit,
                 x='Income',
                 y='Limit',
                 color='Student',
                 opacity=0.8,
                 hover_data=['Student'],
                 title='Income vs Limit',)
fig.update_traces(marker_size=5)
fig.show()

График рассматривает взаимосвязь между доходом и кредитным лимитом теперь не по половому признаку, а по категории в отношении студенчества (учится ли человек в ВУЗе). Видно,что большая часть тех, кто является студентами имеют небольшой доход, соответственно лимит у них тоже не велик. То есть взаисосвязь двух параметров подтверждается.

Нагляднее рассмотреть зависимость кредитного лимита от трех параметров позволит 3д график, для этого разделим на две части (относящихся к студентам и не являющихся таковыми)

In [27]:
non_student = credit[credit.Student == 'No']

In [28]:
fig = px.scatter_3d(non_student, x='Income', y='Balance', z='Limit')
fig.update_traces(marker_size=3, marker_opacity=0.5)
fig.show()

In [29]:
yes_student = credit[credit.Student == 'Yes']

In [30]:
fig = px.scatter_3d(yes_student, x='Income', y='Balance', z='Limit')
fig.update_traces(marker_size=3, marker_opacity=0.5)
fig.show()

Хороший доход, как видно, не всегда гарантирует большой баланс, однако у студентов, которых не так много, оба финансовых параметра не велики, поэтому максимальный кредитный лимит, который они могут получить варьируется в районе 7-9 тысяч.

В то время как у тех, кто не является студентами при их доходах и балансе кредитный лимит достигает 10-13 тысяч.